pkg/volume/metadata/metadata.go

```diff
@@ -0,0 +1,271 @@
+	Name:         "not-used",
+	VolumeSource: api.VolumeSource{EmptyDir: &api.EmptyDirVolumeSource{Medium: api.StorageMediumMemory}},
+}
+
+// readFile reads full path file name and return it as a string
+func readFile(fileName string) string {
+	if data, err := ioutil.ReadFile(fileName); err == nil {
+		return string(data)
+	}
+	return ""
+}
+
+// sortLines sorts the strings generated from map based metadata
+// (annotations and labels)
+func sortLines(values string) string {
+	splitted := strings.Split(values, "\n")
+	sort.Strings(splitted)
+	return strings.Join(splitted, "\n")
+}
+
+// isFileToGenerate compares actual file with the new values. If
+// different (or the file does not exist) return true
+func isFileToGenerate(fileName, values string) bool {
+	if _, err := os.Lstat(fileName); os.IsNotExist(err) {
+		return true
+	}
+	return readFile(fileName) != values
+}
+
+const currentDir = ".current"
+
+// writeDataInTmpDir writes metadata in file
+func (m *metadataVolume) writeDataInTmpDir() (string, error) {
+	errlist := []error{}
+	tempDir, err := ioutil.TempDir(m.GetPath(), "")
+	for fieldReference, fileName := range m.fieldReferenceFileNames {
+		var data string
+		if data, err = fieldpath.ExtractFieldPathAsString(m.pod, fieldReference); err != nil {
+			glog.Error(err)
+			errlist = append(errlist, err)
+		} else {
+			fullPathFile := path.Join(tempDir, fileName)
+			dir, _ := filepath.Split(fullPathFile)
+			if err = os.MkdirAll(dir, os.ModePerm); err != nil {
+				return "", err
+			}
+			if e := ioutil.WriteFile(fullPathFile, []byte(sortLines(data)), 0644); e != nil {
+				glog.Error(e)
+				errlist = append(errlist, err)
+			}
+		}
+	}
+	return tempDir, utilErrors.NewAggregate(errlist)
+}
+
+// updateSymLinks creates the relative links for the all the files
+func (m *metadataVolume) updateSymLinks() error {
+	for _, f := range m.fieldReferenceFileNames {
+		dir, _ := filepath.Split(f)
+		nbOfSubdir := 0
+		if len(dir) > 0 {
+			nbOfSubdir = len(strings.Split(dir, "/")) - 1
+			if e := os.MkdirAll(path.Join(m.GetPath(), dir), os.ModePerm); e != nil {
+				return e
+			}
+		}
+		if _, e := os.Readlink(path.Join(m.GetPath(), f)); e != nil {
+			// link does not exist create it
+			if e := os.Symlink(path.Join(strings.Repeat("../", nbOfSubdir)+currentDir, f), path.Join(m.GetPath(), f)); e != nil {
+				return e
+			}
+		}
+	}
+	return nil
+}
+
+// writeData writes requested metadata in specified files
+func (m *metadataVolume) writeData() error {
+	var tempDir string
+	var err error
+	tempDir, err = m.writeDataInTmpDir()
+	if err != nil {
+		glog.Error(err)
+		return err
+	}
+	_, tempDirBaseName := filepath.Split(tempDir)
+	m.updateSymLinks()
+	var oldTarget string
+	oldTarget, err = os.Readlink(path.Join(m.GetPath(), currentDir))
+	if err == nil {
+		os.Remove(path.Join(m.GetPath(), currentDir)) // remove actual link
+	}
+	err = os.Symlink(tempDirBaseName, path.Join(m.GetPath(), currentDir)) // (re-)link it
```

Comment:
Final layout:
let's assume that the user specify something like this

`metadata.labels` in `labels/data`
`metadata.annotations` in `annotations/data`
`metadata.name` in `podName`

and the volume is mounted on `/opt`

the user will expect to have
`/opt/podName`
`/opt/labels/data`
`/opt/annotations/data`

the plugin creates this (in /opt)
```
podName          -> .current/podName
labels/data      -> .current/labels/data
annotations/data -> .current/annotations/data
.current         ->  058647741 (the tmpDir)
```
`updateSymLinks` func creates the symlinks trees and eventually the directories (in this case: labels and annotations)
while `writeDataInTmpDir` func creates the real data (no symlink):

```
058647741/podName
058647741/labels/data
058647741/annotations/data
```

obviously there's a link between `.current` and the tmpDir

Atomic scheme:
many thanks for the info about atomic update. And many apologizes 'cause this is not the first time you pointed out this. I google a little bit (better and) more and I found what you suggested [here](http://unix.stackexchange.com/a/6786).